deployer/models/strategy.go

```diff
@@ -5,11 +5,90 @@ import (
 	"github.com/coinbase/step/utils/to"
 )
 
+var STRATEGIES = []string{
+	"AllAtOnce",
+	"OneThenAllWithCanary",
+	"25PercentStepRolloutNoCanary",
+	"10PercentStepRolloutNoCanary",
+	"10AtATimeNoCanary",
+	"20AtATimeNoCanary",
+}
+
+type StrategyType string
+
+const (
+	AllAtOnce StrategyType = "AllAtOnce"
+	Canary                 = "Canary"
+	Percent                = "Percent"
+	Increment              = "Increment"
+)
+
+func NewStrategy(autoscaling *AutoScalingConfig, previousDesiredCapacity *int64) *Strategy {
+	// Defaults
+	s := &Strategy{
+		name:                    *autoscaling.Strategy,
+		sType:                   AllAtOnce,
+		minSize:                 int64(1),
+		maxSize:                 int64(1),
+		maxTerminations:         int64(0),
+		spread:                  0.2,
+		previousDesiredCapacity: previousDesiredCapacity,
+	}
+
+	// Get Info from autoscaling
+	if autoscaling.Spread != nil {
+		s.spread = *autoscaling.Spread
+	}
+
+	if autoscaling.MinSize != nil {
+		s.minSize = *autoscaling.MinSize
+	}
+
+	if autoscaling.MaxSize != nil {
+		s.maxSize = *autoscaling.MaxSize
+	}
+
+	if autoscaling.MaxTerminations != nil {
+		s.maxTerminations = *autoscaling.MaxTerminations
+	}
+
+	// Define the Strategy properties
+	switch s.name {
+	case "OneThenAllWithCanary":
+		s.sType = Canary
+	case "25PercentStepRolloutNoCanary":
+		s.sType = Percent
+		// 25% means release is divided into 4 steps
+		s.rollOutSteps = 4
+	case "10PercentStepRolloutNoCanary":
+		s.sType = Percent
+		// 10% means release divided into 10 stages
+		s.rollOutSteps = 10
+	case "10AtATimeNoCanary":
+		s.sType = Increment
+		s.rollOutSteps = float64(s.TargetCapacity()) / float64(10)
+	case "20AtATimeNoCanary":
+		s.sType = Increment
+		s.rollOutSteps = float64(s.TargetCapacity()) / float64(20)
+	}
+
+	return s
+}
+
```

Comment:
for future refactor: would it make sense to have a method like `func stepCount(targetCapacity int) int` on `Strategy` that returned the number of instances to launch at a time as an int given a target capacity value, instead of a float percentage?

Reply:
There are many bad names in this file. I think it is easier to explain percentage, rather than step count, and since these are not descrete "steps" I think the name needs changing. To what I don't know. But since everything in this file is reasonably private at least it is not chaning the API or being used elsewher,e so refactoring is easy.